src/target/riscv/riscv-013.c

```diff
@@ -2837,16 +2837,44 @@ static int deassert_reset(struct target *target)
 
 static int execute_fence(struct target *target)
 {
+	int result;
+	riscv_reg_t saved_mtvec;
+	struct riscv_program program;
+	riscv013_info_t *info = get_info(target);
+
 	if (dm013_select_target(target) != ERROR_OK)
 		return ERROR_FAIL;
 
 	/* FIXME: For non-coherent systems we need to flush the caches right
 	 * here, but there's no ISA-defined way of doing that. */
-	struct riscv_program program;
+
+	/* backup mtvec */
+	result = riscv_get_register(target, &saved_mtvec, GDB_REGNO_MTVEC);
+	if (result != ERROR_OK)
+		return result;
+
+	/* set mtvec after fence.i */
+	result = riscv_set_register(target, GDB_REGNO_MTVEC, info->progbuf_address + 4);
+	if (result != ERROR_OK)
+		return result;
```

Comment:
This isn't great.
This basically loads an address as an instruction into the program buffer (if it is even memory mapped). It is then rewritten when the fence_i is written into the program buffer and executed.